cmd/tlssocksproxy/tlssocksproxy.go

```diff
@@ -6,166 +6,144 @@ import (
-			InsecureSkipVerify: true,
-		}
-		logger.Warn("Running without verification of the tls server - this is dangerous")
+	tlsConfig := &tls.Config{
+		InsecureSkipVerify: *flagInsecureSkipVerify,
 	}
-	ctx := context.Background()
+	if tlsConfig.InsecureSkipVerify {
+		log.Warn("Running without verification of the tls server - this is dangerous")
+	}
+	ctx := cmd.CtxCancelOnOsSignal(log)
 
-	go runPrometheusHandler(ctx, defaultPrometheusAddress)
+	go cmd.RunPrometheusHandler(ctx, log, defaultPrometheusAddress)
 
+	var connID uint64
 	for {
-		socksConn, err := socks5Listener.Accept()
+		localConn, err := localListener.Accept()
 		if err != nil {
-			logger.Fatal(
-				"error accepting incoming connections",
-				zap.Error(err),
-			)
+			log.Fatal("error accepting incoming connections", zap.Error(err))
 		}
-		logger.Info(
-			"socks client connected",
-			zap.String("from", socksConn.RemoteAddr().String()),
-		)
-		go serve(ctx, socksConn, *flagAddrServer, tlsConfig)
+		connID++
+		go serve(ctx, log, localConn, *flagRemoteAddr, tlsConfig, connID)
 	}
 }
 
-func serve(ctx context.Context, srcConn io.ReadWriteCloser, destinationAddress string, tlsConfig *tls.Config) {
-	// Recover if a panic occurs
-	defer recoverAndLogPanic()
-	defer silentClose(srcConn)
-
-	// Cancel context
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
-
+func serve(ctx context.Context, logger *zap.Logger, localConn net.Conn, remoteAddress string, tlsConfig *tls.Config, connID uint64) {
 	start := time.Now()
 
-	dstConn, errDial := tls.DialWithDialer(&net.Dialer{
-		KeepAlive: -1,
-		Timeout:   defaultTimeout,
-	}, "tcp", destinationAddress, tlsConfig)
+	// Recover if a panic occurs
+	defer cmd.RecoverAndLogPanic(logger)
+	defer cmd.SilentClose(localConn)
 
-	if errDial != nil {
-		logger.Warn(
-			"could not reach tls server",
-			zap.Error(errDial),
-		)
+	remoteConn, err := tls.DialWithDialer(&net.Dialer{
+		Timeout: defaultTimeout,
+	}, "tcp", remoteAddress, tlsConfig)
+	if err != nil {
+		logger.Warn("could not reach remote tls server", zap.Error(err))
 		return
 	}
-	defer silentClose(dstConn)
-
-	group, gctx := errgroup.WithContext(ctx)
-
-	group.Go(func() error {
-		srcConn := tlssocks.NewBufferedReader(gctx, srcConn)
-		return copyData("conn->socksConn", dstConn, srcConn)
-	})
-	group.Go(func() error {
-		dstConn := tlssocks.NewBufferedReader(gctx, dstConn)
-		return copyData("socksConn->conn", srcConn, dstConn)
-	})
-
-	if err := group.Wait(); err != nil {
-		switch {
-		case err == io.ErrUnexpectedEOF,
-			err == io.ErrClosedPipe,
-			err == io.EOF,
-			err.Error() == "broken pipe":
-			logger.Warn("Error occurred, while copying data", zap.Error(err))
-		default:
-			logger.Error("Unexpected error occurred while copying the data", zap.Error(err))
-		}
+	defer cmd.SilentClose(remoteConn)
+
+	p := &proxy{
+		log:  logger,
+		wait: make(chan struct{}),
 	}
 
+	deadline := start.Add(connDeadline)
+	_ = localConn.SetDeadline(deadline)
```

Comment:
@smartinov 
Given the usage of the code, is the 1 minute timeout enough to cover the use cases?